staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -1338,3 +1460,122 @@ func buildOpenAPIModelsForApply(staticOpenAPISpec *goopenapispec.Swagger, crd *a
 	}
 	return models, nil
 }
+
+func (r *crdHandler) updateStorageVersionFor(crd *apiextensionsv1.CustomResourceDefinition) error {
+	gr := schema.GroupResource{
+		Group:    crd.Spec.Group,
+		Resource: crd.Spec.Names.Plural,
+	}
+	storageVersion, err := apiextensionshelpers.GetCRDStorageVersion(crd)
+	if err != nil {
+		// This should never happened if crd is valid, which is true since we
+		// only update storage version for CRDs that have been written to the
+		// storage
+		return err
+	}
+	encodingVersion := crd.Spec.Group + "/" + storageVersion
+	var decodableVersions []string
+	for _, v := range crd.Spec.Versions {
+		decodableVersions = append(decodableVersions, crd.Spec.Group+"/"+v.Name)
+	}
+
+	appendOwnerRefFunc := func(sv *apiserverinternalv1alpha1.StorageVersion) error {
+		ref := metav1.OwnerReference{
+			APIVersion: apiextensionsv1.SchemeGroupVersion.String(),
+			Kind:       "CustomResourceDefinition",
+			Name:       crd.Name,
+			UID:        crd.UID,
+		}
+		for _, r := range sv.OwnerReferences {
+			if r == ref {
+				return nil
+			}
+		}
+		sv.OwnerReferences = append(sv.OwnerReferences, ref)
+		return nil
+	}
+	return storageversion.UpdateStorageVersionFor(
+		r.storageVersionClient,
+		r.apiserverID,
+		gr,
+		encodingVersion,
+		decodableVersions,
+		appendOwnerRefFunc)
+}
+
+// getOrCreateStorageVersionUpdateQueue_locked returns the channel for the given UID,
+// or create a new one and a new goroutine if necessary. The goroutine keeps processing
+// requests until the channel is closed.
+// NOTE: caller should hold the customStorageLock until the write to the returned channel is finished,
+// to make sure the storage version update is in sync with the underlying storage change.
+func (r *crdHandler) getOrCreateStorageVersionUpdateQueue_locked(uid types.UID) chan *storageVersionUpdateRequest {
+	if queue, ok := r.storageVersionUpdateQueues[uid]; ok {
+		return queue
+	}
+
+	queue := make(chan *storageVersionUpdateRequest, maxStorageVersionUpdateRequests)
+	r.storageVersionUpdateQueues[uid] = queue
+	go func() {
+		for req := range queue {
+			if req.wait != nil {
+				<-req.wait
+			}
+			// There is a chance that if:
+			//   1. the CRD gets deleted and re-created in a short time, and
+			//      the storage version object hasn't been GC'ed yet
+			//   2. the old queue gets drained, but one req is still being processed
+			//   3. the re-created CRD establishes a new queue and triggers a
+			//      new storage version update
+			// the old storage version update request can race against the new
+			// queue, and causing the storage version to go back in time.
+			// TODO(roycaihw): find a way to cancel the req when we drain the
+			// old queue (e.g. use context)
```

Comment:
I think we should add a cancel() function to the req, and call cancel() for all remaining reqs when stopStorageVersionUpdateQueue_locked is called, then we don't have this race.

Reply:
Another use of the cancel() function is that when a new req is enqueued, we can cancel the existing storage version updates (and fail the pending CR write requests). Note that we still need to wait for pending teardown to finish draining.

Reply:
addressed the TODOs

Reply:
> add a cancel() function to the req

note that the cancel() function shouldn't be attached to a specific req, since the race here is about not being able to grab a req. I made all the req in one queue share the same context, and attach the cancel() function to the queue. 

> we can cancel the existing storage version updates (and fail the pending CR write requests)

I'm not fond of failing CR requests when we don't have to. The requests should wait for the storage version update and then proceed. The only place we have to fail in-flight requests is when a CRD is deleted-- we have to tear down the queue to avoid leaking goroutine, and we have to do it in a quick manner because we don't know if the CRD will be re-created and trigger new storage version updates.